unliftio/src/UnliftIO/Async.hs

```diff
@@ -308,6 +327,247 @@ replicateConcurrently i m = withRunInIO $ \run -> A.replicateConcurrently i (run
+
+-- | Similar to 'pooledMapConcurrentlyN' but with number of threads
+-- set from 'getNumCapabilities'. Usually this is useful for CPU bound
+-- tasks.
+--
+-- @since 0.2.9
+pooledMapConcurrently :: (MonadUnliftIO m, Traversable t) => (a -> m b) -> t a -> m (t b)
+pooledMapConcurrently f xs = do
+  withRunInIO $ \run -> do
+    numProcs <- getNumCapabilities
+    pooledMapConcurrentlyIO numProcs (run . f) xs
+
+-- | Similar to 'pooledMapConcurrentlyN' but with flipped arguments.
+--
+-- @since 0.2.9
+pooledForConcurrentlyN :: (MonadUnliftIO m, Traversable t) 
+                      => Int -- ^ Max. number of threads. Should not be less than 1.
+                      -> t a -> (a -> m b) -> m (t b)
+pooledForConcurrentlyN numProcs xs f = flip (pooledMapConcurrentlyN numProcs) xs f
+
+-- | Similar to 'pooledForConcurrentlyN' but with number of threads
+-- set from 'getNumCapabilities'. Usually this is useful for CPU bound
+-- tasks.
+--
+-- @since 0.2.9
+pooledForConcurrently :: (MonadUnliftIO m, Traversable t) => t a -> (a -> m b) -> m (t b)
+pooledForConcurrently xs f = flip pooledMapConcurrently xs f
+
+pooledMapConcurrentlyIO :: Traversable t => Int -> (a -> IO b) -> t a -> IO (t b)
+pooledMapConcurrentlyIO numProcs f xs = 
+    if (numProcs < 1)
+    then error "pooledMapconcurrentlyIO: number of threads < 1"
+    else pooledMapConcurrentlyIO' numProcs f xs
+
+pooledMapConcurrentlyIO' ::
+  Traversable t => Int -> (a -> IO b) -> t a -> IO (t b)
+pooledMapConcurrentlyIO' numProcs f xs = do
+  -- prepare one IORef per result...
+  jobs :: t (a, IORef b) <-
+    for xs (\x -> (x, ) <$> newIORef (error "pooledMapConcurrentlyIO': empty IORef"))
+  -- ...put all the inputs in a queue..
+  jobsVar :: MVar [(a, IORef b)] <- newMVar (toList jobs)
+  -- ...run `numProcs` threads in parallel, each
+  -- of them consuming the queue and filling in
+  -- the respective IORefs.
+  forConcurrently_ [1..numProcs] $ \_ -> do
+    let loop  = do
+          mbJob :: Maybe (a, IORef b) <- modifyMVar jobsVar $ \x -> case x of
+            [] -> return ([], Nothing)
+            var : vars -> return (vars, Just var)
+          case mbJob of
+            Nothing -> return ()
+            Just (x, outRef) -> do
+              y <- f x
+              writeIORef outRef y
+              loop
+    loop
+  -- Read all the IORefs
+  for jobs (\(_, outputRef) -> readIORef outputRef)
+
+pooledMapConcurrentlyIO_' ::
+  Traversable t => Int -> (a -> IO b) -> t a -> IO ()
+pooledMapConcurrentlyIO_' numProcs f jobs = do
+  jobsVar :: MVar [a] <- newMVar (toList jobs)
+  forConcurrently_ [1..numProcs] $ \_ -> do
+    let loop  = do
+          mbJob :: Maybe a <- modifyMVar jobsVar $ \x -> case x of
+            [] -> return ([], Nothing)
+            var : vars -> return (vars, Just var)
+          case mbJob of
+            Nothing -> return ()
+            Just x -> do
+              y <- f x
+              loop
+    loop
+  return ()
```

Comment:
There is a lot of code duplication in above two functions, consider abstracting the insides out:
```haskell
pooledMapConcurrentlyIO' ::
  Traversable t => Int -> (a -> IO b) -> t a -> IO (t b)
pooledMapConcurrentlyIO' numProcs f xs = do
  -- prepare one IORef per result...
  jobs :: t (a, IORef b) <-
    for xs (\x -> (x, ) <$> newIORef (error "pooledMapConcurrentlyIO': empty IORef"))
  -- ...put all the inputs in a queue..
  jobsVar :: MVar [(a, IORef b)] <- newMVar (toList jobs)
  -- ...run `numProcs` threads in parallel, each
  -- of them consuming the queue and filling in
  -- the respective IORefs.
  pooledConcurrently numProcs jobsVar $ \ (x, outRef) -> f x >>= writeIORef outRef
  -- Read all the IORefs
  for jobs (\(_, outputRef) -> readIORef outputRef)

pooledMapConcurrentlyIO_' ::
  Traversable t => Int -> (a -> IO b) -> t a -> IO ()
pooledMapConcurrentlyIO_' numProcs f jobs = do
  jobsVar :: MVar [a] <- newMVar (toList jobs)
  pooledConcurrently numProcs jobsVar f

pooledConcurrently
  :: Int -> MVar [a] -> (a -> IO b) -> IO ()
pooledConcurrently numProcs jobsVar f = do
  forConcurrently_ [1..numProcs] $ \_ -> do
    let loop  = do
          mbJob :: Maybe a <- modifyMVar jobsVar $ \x -> case x of
            [] -> return ([], Nothing)
            var : vars -> return (vars, Just var)
          case mbJob of
            Nothing -> return ()
            Just x -> do
              _ <- f x
              loop
    loop
```

Reply:
Done.